File: src/main/java/com/github/simiacryptus/aicoder/openai/translate/TranslationRequest_XML.java
package com.github.simiacryptus.aicoder.openai.translate;

import com.github.simiacryptus.aicoder.config.AppSettingsState;
import com.github.simiacryptus.aicoder.openai.CompletionRequest;
import org.jetbrains.annotations.NotNull;

import java.util.stream.Collectors;

public class TranslationRequest_XML extends BaseTranslationRequest<TranslationRequest_XML> {

    public TranslationRequest_XML(@NotNull AppSettingsState settings) {
        setTemperature(settings.temperature);
        setMaxTokens(settings.maxTokens);
    }

    @Override
    @NotNull
    public CompletionRequest buildCompletionRequest() {
        @NotNull CharSequence inputAttrStr = getInputAttr().isEmpty() ? "" : (" " + getInputAttr().entrySet().stream().map(t -> String.format("%s=\"%s\"", t.getKey(), t.getValue())).collect(Collectors.joining(" ")));
        @NotNull CharSequence outputAttrStr = getOutputAttr().isEmpty() ? "" : (" " + getOutputAttr().entrySet().stream().map(t1 -> String.format("%s=\"%s\"", t1.getKey(), t1.getValue())).collect(Collectors.joining(" ")));
        @NotNull String text = String.format("<!-- %s -->\n<%s%s>%s</%s>\n<%s%s>",
                getInstruction(),
                getInputTag().toLowerCase(),
                inputAttrStr,
                getOriginalText().trim(),
                getInputTag().toLowerCase(),
                getOutputTag().toLowerCase(),
                outputAttrStr
        ).trim();
        return new CompletionRequest(
                text,
                getTemperature(),
                getMaxTokens(),
                null,
                String.format("</%s>", getOutputTag().toLowerCase())
        );
    }

}

File: src/main/kotlin/com/github/simiacryptus/aicoder/openai/translate/TranslationRequest_XML.kt
package com.github.simiacryptus.aicoder.openai.translate

import com.github.simiacryptus.aicoder.config.AppSettingsState
import com.github.simiacryptus.openai.CompletionRequest
import java.util.*
import java.util.stream.Collectors

class TranslationRequest_XML(settings: AppSettingsState) :
    BaseTranslationRequest<TranslationRequest_XML>(settings) {

    override fun buildCompletionRequest(): CompletionRequest {
        val inputAttrStr: CharSequence = if (inputAttr.isEmpty()) "" else " " + inputAttr.entries.stream()
            .map { (key, value) -> """$key="$value"""" }.collect(Collectors.joining(" "))
        val outputAttrStr: CharSequence = if (outputAttr.isEmpty()) "" else " " + outputAttr.entries.stream()
            .map { (key, value) -> """$key="$value"""" }.collect(Collectors.joining(" "))
        val exampleStr: CharSequence =
            if (examples.isEmpty()) "" else ("<examples>" + examples.joinToString("\n\t") + "</examples>\n")
        val inputTagTxt = inputTag.toString().lowercase(Locale.getDefault())
        val outputTagTxt = outputTag.toString().lowercase(Locale.getDefault())
        val inputText = originalText.toString().trim { it <= ' ' }
        return CompletionRequest(
            """
            <!-- $instruction -->
            <$inputTagTxt$inputAttrStr>$inputText</$inputTagTxt>
            $exampleStr
            <$outputTagTxt$outputAttrStr>
            """.trimIndent().trim(),
            temperature,
            maxTokens,
            null,
            """</$outputTagTxt>"""
        )
    }

    private val examples: MutableList<CharSequence> = ArrayList()
    override fun addExample(
        exampleText: CharSequence,
        attributes: Map<CharSequence, CharSequence>
    ): TranslationRequest {
        val outputTagTxt = outputTag.toString().lowercase(Locale.getDefault())
        val outputAttrStr = if (attributes.isEmpty()) "" else " " + attributes.entries.stream()
            .map { (key, value) -> "$key=\"$value\"" }.collect(Collectors.joining(" "))
        examples.add("""<$outputTagTxt$outputAttrStr>$exampleText</$outputTagTxt>""")
        return this
    }
}